### PR TITLE
feat: validate unparsed AsyncAPI document

### PR DESCRIPTION
**Pull request by magicmatatjahu**
**Description**

Validate unparsed AsyncAPI document (before `$ref`s resolving).

### CODE REVIEW

**Comment by magicmatatjahu** [comment]
@jonaslagoni Could you check?

File: test/ruleset/rules/asyncapi-document-unresolved.spec.ts [hunk 1]
```diff
@@ -1,90 +1,85 @@
...
+      },
+      components: {
+        'x-operations': {
+          someOperation: {},
+        },
+      },
+    },
+    errors: [
+      {
+        message: 'Referencing in this place is not allowed',
```
Comment (jonaslagoni) [inline test/ruleset/rules/asyncapi-document-unresolved.spec.ts:79]: Why is this needed? Is it not caught by the spec-schema?

File: test/ruleset/rules/asyncapi-document-unresolved.spec.ts [hunk 2]
```diff
@@ -1,90 +1,85 @@
...
```
Comment (jonaslagoni) [inline test/ruleset/rules/asyncapi-document-unresolved.spec.ts:53]: Why is this needed? Is it not caught by the spec-schema?

**Review by jonaslagoni — CHANGES_REQUESTED** [review]
I guess I dont understand why we don't change the parsing logic to validate before resolving references, and then again after? This fix looks like an incorrect approach to take 🤔 Or is there some underlying reason?

Related to https://github.com/asyncapi/parser-js/issues/405

**Comment by magicmatatjahu** [comment]
@jonaslagoni 

> I guess I dont understand why we don't change the parsing logic to validate before resolving references, and then again after? This fix looks like an incorrect approach to take 🤔 Or is there some underlying reason?

Spectral resolver (and other json-schema resolvers) doesn't validate the json-schema but only resolve it. Also AJV try to validate the json-schema based on json-schema and that's the problem. After referencing we don't know if `$ref` in given place can be located and/or given ref has valid object (for given location, e.g. message etc.). So we exactly validate two things: unresolved AsyncAPI document with refs and with resolved refs (even that deep). I don't know how to make it in different way. That PR is related to the https://github.com/asyncapi/parser-js/issues/405

**Comment by jonaslagoni** [comment]
Ahh shit, validation happens through spectral now, yea then you cant do a simple double validation once before refs are resolved and once after...

**Comment by magicmatatjahu** [comment]
@jonaslagoni 

> Ahh shit, validation happens through spectral now, yea then you can't do a simple double validation once before refs are resolved and once after...

this is what I do 😄

**Review by jonaslagoni — APPROVED** [review]

**Pull request merged**